elytron/WFLY-16306-keycloak-saml-feature-pack.adoc

```diff
@@ -0,0 +1,331 @@
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/EAP7-1928[EAP7-1928]
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+=== Testing By
+
+TBD.
+
+Note that since this feature is being developed across two different projects, we'll need engineers and QE
+from both the WildFly and Keycloak projects involved.
+
+* [ ] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+ * https://github.com/keycloak/keycloak
+ * https://github.com/wildfly
+
+=== Other Interested Projects
+
+== Requirements
+
+=== Hard Requirements
+
+Note that the code for the Keycloak Elytron SAML adapter is going to be extracted from the https://github.com/keycloak/keycloak repo
+and moved to a separate repo. We don't have details yet on when this will happen. We'll likely need to add our feature pack to the
+existing Keycloak project initially and then it will get moved to the new project with the rest of the adapter code.
+
+A new Galleon feature pack will be added to the Keycloak project with the group ID, `org.keycloak`, and artifact ID, `keycloak-saml-adapter-galleon-pack`.
+This new feature pack will contain three layers called `keycloak-saml`, `keycloak-client-saml`, and `keycloak-client-saml-ejb`.
+
+When this feature pack and its layers are provisioned to a WildFly installation, the end result will mostly match the results
+of manually installing the Keycloak Elytron SAML adapter using the instructions linked above.
+
+In particular, when the `keycloak-saml` layer is provisioned, the following steps will occur:
+
+ # Install a set of modules under `modules/system/add-ons`
+ # Register the `org.keycloak.keycloak-saml-adapter-subsystem` extension.
+ # Add an empty `urn:jboss:domain:keycloak:1.4` subsystem.
+
+The `keycloak-client-saml` layer will depend on the `keycloak-saml` layer and when provisioned, the following additional steps will occur:
+
+ # Add the required resources to the `elytron` subsystem to result in an `http-authentication-factory` and `security-domain` supporting SAML being available.
+ # Update the `application-security-domain` mapping in the `undertow` subsystem to map the default security domain to the new `http-authentication-factory`.
+
+The `keycloak-client-saml-ejb` layer will depend on the `keycloak-client-saml` layer and when provisioned, the following additional steps will occur:
+
+ # Update the `application-security-domain` mapping in the `ejb3` subsystem to map the default security domain to the new `KeycloakDomain`.
+
+Note that the existing CLI scripts for the SAML adapter and the existing feature pack for OIDC only update the `application-security-domain` mapping in the
+`undertow` subsystem. However, the CLI scripts for the OIDC adapter also update the `application-security-domain`
+mapping in the EJB3 subsystem. For this reason, we have added the `keycloak-client-saml-ejb` layer here for users
+that need to propagate their identities to EJBs as well.
```

Comment:
@fjuma it's not clear to me if we have 3 layers for OIDS too, I mean e.g. `keycloak-oidc`, `keycloak-client-oidc`, `keycloak-client-oidc-ejb`

Reply:
@tommaso-borgato For OIDC, we are no longer using the Keycloak OIDC adapter. WildFly's native support for OIDC is used instead as described here:

https://www.wildfly.org/news/2021/10/12/bootable-jar-6.0.Final-Released/


https://github.com/wildfly/wildfly/blob/main/docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Elytron_OIDC_Client.adoc#L79

The new Galleon layer is called `elytron-oidc-client`.